### ISSUE TITLE
Load learner cannot find models when not running in pythons __main__

### ISSUE THREAD

**Issue by Baukebrenninkmeijer** [fastai/fastai#2575]
**Describe the bug**
I'm trying to run a model using flask. Initially I had the flask setup as follows:
```
├── src
│    ├── __init__.py
│    ├── models.py
│    └── utils.py
├── app
│    ├── __init__.py  # instantiates flask app
│    └── routes.py  # imports the models from src.models
├── main.py -> Runs the flask app
```
Now, as far as I know, this is a fairly standard flask setup, used in many tutorials as well. But no matter where I put my model definition, it always gives the error 
```
AttributeError: Can't get attribute 'ModelResnetClassifier' on <module '__main__' from 'main.py'
````
To fix this issue, I expected putting everything in `main.py` would fix it, looking at issues like #2436 where this was mentioned as a possible reason. But the issue goes a bit deeper than just being in the same script, because that is what I'm doing now. The issue persists almost anytime that `load_learner` is not being called in '__main__'. 

I am now trying to run this a bit faster using gunicorn. Since now the '__main__' is changed again to gunicorns wrapper, the same error is occuring.
```
AttributeError: Can't get attribute 'ModelResnetClassifier' on <module '__main__' from '/usr/local/bin/gunicorn'>
```
I've spent countless hours by now trying to fix the issue, with moving the model definition, changing names, setting the model as attribute of '__main__', and more. But so far, the only way to save things is and load them without this issue is to use predefined models from libraries, that have their model definition there. 

**Provide your installation details**
```text
=== Software === 
python        : 3.7.6
fastai        : 1.0.60
fastprogress  : 0.2.2
torch         : 1.3.1
nvidia driver : 445.75
torch cuda    : 10.1 / is available
torch cudnn   : 7501 / is enabled

=== Hardware === 
nvidia gpus   : 1
torch devices : 1
  - gpu0      : 6144MB | GeForce GTX 980 Ti

=== Environment === 
platform      : Windows-10-10.0.18362-SP0
conda env     : Unknown
python        : C:\Users\BaukeBrenninkmeijer\Anaconda3\envs\otm\python.exe
sys.path      : C:\Users\BaukeBrenninkmeijer\Projects\OneTwoModel\nbs
C:\Users\BaukeBrenninkmeijer\Anaconda3\envs\otm\python37.zip
C:\Users\BaukeBrenninkmeijer\Anaconda3\envs\otm\DLLs
C:\Users\BaukeBrenninkmeijer\Anaconda3\envs\otm\lib
C:\Users\BaukeBrenninkmeijer\Anaconda3\envs\otm

C:\Users\BaukeBrenninkmeijer\Anaconda3\envs\otm\lib\site-packages
C:\Users\BaukeBrenninkmeijer\Anaconda3\envs\otm\lib\site-packages\win32
C:\Users\BaukeBrenninkmeijer\Anaconda3\envs\otm\lib\site-packages\win32\lib
C:\Users\BaukeBrenninkmeijer\Anaconda3\envs\otm\lib\site-packages\Pythonwin
C:\Users\BaukeBrenninkmeijer\Anaconda3\envs\otm\lib\site-packages\IPython\extensions
C:\Users\BaukeBrenninkmeijer\.ipython
```

**To Reproduce**
1. Create folder structure like described:
```
├── src
│    └── models.py
├── app
│    ├── __init__.py  # instantiates flask app
│    └── routes.py  # imports the models from src.models
├── main.py -> Runs the flask app
```
2. Try to run a custom model (either defined in routes or src.models), loaded with `load_learner` from `routes.py`. 

**Expected behavior**
That `load_learner` is able to import the model if it is anywhere in the available namespace or an easy way to match the requirements that `load_learner` has for these models. 

**Screenshots**

![image](https://user-images.githubusercontent.com/9077462/81785768-fe766f00-94fe-11ea-8d5d-0bc3822bfea4.png)

**Additional context**

**Comment by sgugger** [comment]
This has already been addressed before. `load_learner` serializes function names using pickle, so you need to use the **same module** at training at inference time containing all your extra stuff.

Here `load_learner` is looking for `ModelResnetClassifier` in main because it was defined there when you exported the `Learner`. You need to fix your export file by exporting your Learner in a script that will have `ModelResnetClassifier` come from the same python modules you are using in production.

**Comment by Baukebrenninkmeijer** [comment]
Hmm alright, I see. Let me try that! Thanks!

**Comment by Perseus14** [comment]
@Baukebrenninkmeijer I am facing the same issue, were you able to fix it?

**Comment by Baukebrenninkmeijer** [comment]
Yes, if I remember correctly, you export the model to a .py, import that into your notebooks or wherever you are saving the model. Then, during inference you load from that same .py file, and it should work. The reference to the object then matches in both cases.

**Comment by AdamZWinter** [comment]
I am having the same problem, but not sure I follow what you are saying to do.  Can you provide and example?  Thank you.

**Comment by Baukebrenninkmeijer** [comment]
@AdamZWinter It's been a while, but let me try. If I recall correctly, the issue was I was training something in a notebook, which always has the namespace `__main__`. I then saved the model into a pickle. Upon loading the model, it expects to have the model definition again in `__main__`, but I copied it to a module to be used in the API. The fix was to have both training and inference use the same model in the same module (i.e. `models.py` for example).
